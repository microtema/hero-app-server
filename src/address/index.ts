import {Address} from './Address';
import AddressRepository from './Address.repository';
import AddressService from './Address.service';

export {Address, AddressService, AddressRepository};
