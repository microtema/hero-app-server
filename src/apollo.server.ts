import {ApolloServer as ApolloServerExpress, gql} from 'apollo-server-express';
import {Inject, Singleton} from 'typescript-ioc';
import AuthorService from './author/AuthorService';
import BookService from './book/BookService';

@Singleton
export default class ApolloServer {

    public readonly server: ApolloServerExpress;

    public readonly resolvers = {
        Query: {
            author(parent, {id}, {authorService}, info) {

                return authorService.getAuthor(id);
            },
            authors(parent, {name}, {authorService}, info) {

                return authorService.getAuthors({name});
            },
            book(parent, {id}, {bookService}, info) {

                return bookService.getBook(id);
            },
            books(parent, {title}, {bookService}, info) {

                return bookService.getBooks({title});
            },
        },

        Mutation: {
            createAuthor(parent, {name}, {authorService}, info) {

                return authorService.createAuthor({name});
            },

            updateAuthor(parent, {id, name}, {authorService}, info) {

                return authorService.updateAuthor({id, name});
            },

            deleteAuthor(parent, {id}, {authorService}, info) {

                return authorService.deleteAuthor(id);
            },

            createBook(parent, {title, authorId}, {bookService}, info) {

                return bookService.createBook({title, authorId});
            },

            updateBook(parent, {id, title}, {bookService}, info) {

                return bookService.updateBook({id, title});
            },

            deleteBook(parent, {id}, {bookService}, info) {

                return bookService.deleteBook(id);
            },
        },
    };

    // Create GPL Schema
    public readonly typeDefs = gql`
        type Query {
            author(id: ID): Author
            authors(name: String): [Author]
            book (id: ID): Book
            books (title: String): [Book]
        }

        type Mutation {
            createAuthor(name: String): Author
            updateAuthor(id: ID,name: String): Author
            deleteAuthor(id: ID): Boolean
            createBook(title: String, authorId:ID): Book
            updateBook(id: ID, title: String): Book
            deleteBook(id: ID): Boolean
        }

        type Book {
            id: ID
            title: String
            author: Author
        }

        type Author {
            id: ID
            name: String
            books: [Book]
        }
    `;

    public readonly config = {
        context: ({req}) => ({
            authorService: this.authorService,
            bookService: this.bookService,
            req,
        }), // pass in the request and services to keep resolvers cleaner
        resolvers: this.resolvers,
        typeDefs: this.typeDefs,
    };

    constructor(@Inject private authorService: AuthorService, @Inject private bookService: BookService) {

        this.server = new ApolloServerExpress(this.config);
    }
}
