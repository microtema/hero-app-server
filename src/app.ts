import express from 'express';

const app = express();

app.get('/', (req, res) => {

    res.send('The sedulous hyena ate the antelope!');
});

export default app;
