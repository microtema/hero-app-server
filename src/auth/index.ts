import AuthService from './Auth.service';

export {AuthService};
