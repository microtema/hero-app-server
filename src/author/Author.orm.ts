const AuthorModel = (sequelize, DataTypes) => {

    const model = sequelize.define('author', {
        name: {
            type: DataTypes.STRING,
        },
    });

    model.associate = (models) => {
        model.hasMany(models.BookModel, {onDelete: 'CASCADE'});
    };

    return model;
};

export default AuthorModel;
