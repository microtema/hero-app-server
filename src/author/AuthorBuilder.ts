import faker from 'faker';
import {ModelBuilder} from 'model-builder-ts';
import {Inject, Singleton} from 'typescript-ioc';
import BookBuilder from '../book/BookBuilder';
import {Author} from './Author';

@Singleton
export default class AuthorBuilder extends ModelBuilder<Author> {

    constructor(@Inject private bookBuilder: BookBuilder) {
        super();
    }

    public min(): Author {

        const author = {
            books: this.bookBuilder.list(3),
            id: faker.random.uuid(),
            name: faker.name.findName(),
        } as Author;

        author.books.forEach((it) => it.author = author);

        return author;
    }
}
