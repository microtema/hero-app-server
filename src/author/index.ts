import AuthorRepository from './AuthorRepository';
import AuthorService from './AuthorService';

export {AuthorRepository, AuthorService};
