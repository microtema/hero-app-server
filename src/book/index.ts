import {Book} from './Book';
import BookRepository from './Book.repository';
import BookService from './Book.service';

export {BookRepository, BookService, Book};
