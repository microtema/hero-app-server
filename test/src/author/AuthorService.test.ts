import {AuthorRepository, AuthorService} from '../../../src/author';
import {BookBuilder} from '../../../src/book';
import AuthorBuilder from './AuthorBuilder';

describe('Test AuthorService', () => {

    const authorRepository = {
        delete: jest.fn(() => Promise.resolve({})) as any,
        findAll: jest.fn(() => Promise.resolve({})) as any,
        findByPk: jest.fn(() => Promise.resolve({})) as any,
        save: jest.fn(() => Promise.resolve({})) as any,
        update: jest.fn(() => Promise.resolve({})) as any,
    } as AuthorRepository;

    const sut = new AuthorService(authorRepository);
    const builder = new AuthorBuilder(new BookBuilder());

    it('It should be defined', () => {

        expect(sut).toBeDefined();
    });

    it('getAuthors filter by name', () => {

        // given
        const {name} = builder.min();

        // when
        const promise = sut.getAuthors({name});

        // then
        expect(promise).toBeDefined();
        expect(authorRepository.findAll).toBeCalledWith(name);
    });

    it('getAuthor by id', () => {

        // given
        const {id} = builder.min();

        // when
        const promise = sut.getAuthor(id);

        // then
        expect(promise).toBeDefined();
        expect(authorRepository.findByPk).toBeCalledWith(id);
    });

    it('deleteAuthor by id', () => {

        // given
        const {id} = builder.min();

        // when
        const promise = sut.deleteAuthor(id);

        // then
        expect(promise).toBeDefined();
        expect(authorRepository.delete).toBeCalledWith(id);
    });

    it('createAuthor', () => {

        // given
        const author = builder.min();

        // when
        const promise = sut.createAuthor(author);

        // then
        expect(promise).toBeDefined();
        expect(authorRepository.save).toBeCalledWith(author);
    });

    it('updateAuthor', () => {

        // given
        const author = builder.min();

        // when
        const promise = sut.updateAuthor(author);

        // then
        expect(promise).toBeDefined();
        expect(authorRepository.update).toBeCalledWith(author);
    });
});
