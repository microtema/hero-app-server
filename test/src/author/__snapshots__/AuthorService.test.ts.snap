// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AuthorService getAuthor by id should  not match 1`] = `undefined`;

exports[`Test AuthorService getAuthor by id should match 1`] = `undefined`;

exports[`Test AuthorService getAuthors by name should match 1`] = `undefined`;

exports[`Test AuthorService getAuthors by name should not match 1`] = `undefined`;
